Docs: Explaining plugin machinery for contributors (#2582)
flyteorg/flyte: rsts/concepts/lifecycle_workflow_with_plugin.rst

> .. _lifecycle-workflow:

Reviewer: ```suggestion
.. _workflow-lifecycle:
```

Reviewer: Could you rename the file to workflow_lifecycle.rst?

#################################################################
Understand the Lifecycle of a Flyte Workflow
#################################################################

After reading this entry you should be able to understand how Flyte's Plugin machinery
work under the hood and how information flows from the different
components in Flyte.

> 

Reviewer 2: Let's understand how Flyte's Plugin machinery works and how information flows from one component to another in Flyte.

> Under the hood Flyte relies on a primitive called “Plugins”. Every task

Reviewer 2: Under the hood,

that you run on Flyte is somehow powered by a plugin. Some of those plugins are native and guaranteed by Flyte system. These native plugins are for example powering your python functions (i.e: Tasks) by running them inside a k8s pod. There are Three native plugins : ``Container``, ``K8sPod`` and ``Sql``.

> 

Reviewer 2: Under the hood, Flyte relies on a primitive called “Plugins”. Every task that you run on Flyte is powered by a plugin. Some of these plugins are native and guaranteed by Flyte system. These native plugins, for example, run your Python functions (such as tasks) inside a k8s pod. There are three native plugins: ``Container``, ``K8sPod`` and ``Sql``.

Addionally there are also plugins which are actual extensions, they create additional infrastructure, talk to other SaaS on your behalf. Examples of those are the Spark or AWS Athena plugins.

> A plugin requires code to live in multiple locations

Reviewer: @dav009, could you explain your intention here? I don't understand this statement.

Author: @samhita-alla  I want to convey that a plugin code most likely implies some code living in the operator and some code living in the SDK, both of which are different repositories and/or programming languages.

> 

Reviewer 2: There are plugins which are actual extensions; they create additional infrastructure, and communicate with SaaS on your behalf. Example: Spark, AWS Athena.

 A plugin requires code to live in multiple locations

Reviewer 2: Can we add links to Spark and AWS Athena plugin?

1. Some plugin logic has to live in SDK so that users can create tasks
   that make use of a plugin. This logic usually goes into a specific
   SDK i.e: flytekit’s python: https://github.com/flyteorg/flyteplugins

> 

Reviewer 2: Plugin logic lives in the SDK for users to create tasks that uses the plugin. This logic resides in the goes specific SDK, that is: flytekit’s python: https://github.com/flyteorg/flyteplugins.

Reviewer: Are you referring to https://github.com/flyteorg/flytekit/tree/master/plugins? I don't think the link that has been attached is correct.

Author: @samhita-alla you are right, we should link  ` https://github.com/flyteorg/flytekit/tree/master/plugins`.

2. Another big chunk of plugins logic lives in
   `Flyteplugins <https://github.com/flyteorg/flyteplugins>`__. This
   repository is a library that gets loaded into
   `FlytePropeller <https://github.com/flyteorg/flytepropeller>`__.

>    Flytepropeller is a Kubernetes operator, when Flytepropeller starts

Reviewer 2: This is a library that is loaded into `FlytePropeller <https://github.com/flyteorg/flytepropeller>`__.

>    Flytepropeller is a Kubernetes operator, when Flytepropeller starts
>    it loads Flyteplugins. Flytepropeller knows that plugins exists and that
>    executing tasks depends on them. However Flytepropeller knows nothing
>    about “how to execute plugins”.

Reviewer: ```suggestion
   FlytePropeller (a Kubernetes operator) loads Flyteplugin upon starting. 
   Flytepropeller knows that plugins exist and that
   executing tasks depends on them. However, Flytepropeller knows nothing
   about “how to execute the plugins”.
```

> 

Reviewer 2: FlytePropeller (a Kubernetes operator) loads Flyteplugin upon starting.
